src/mesh/topology.rs

```diff
@@ -64,27 +88,25 @@ pub fn compute_face_to_face_topology(mesh: &Mesh) -> FaceToFaceTopology {
 /// A topology containing neighborhood relations between vertices. Two
 /// vertices are neighbors if and only if they they are end points of
 /// an edge.
-pub type VertexToVertexTopology = HashMap<u32, SmallVec<[u32; MAX_INLINE_NEIGHBOR_COUNT]>>;
+pub type VertexToVertexRelation = SmallVec<[u32; MAX_INLINE_NEIGHBOR_COUNT]>;
 
 /// Computes vertex to vertex topology for mesh.
-pub fn compute_vertex_to_vertex_topology(mesh: &Mesh) -> VertexToVertexTopology {
-    let mut v2v: HashMap<u32, SmallVec<[u32; 8]>> = HashMap::new();
+pub fn compute_vertex_to_vertex_topology(mesh: &Mesh) -> Vec<VertexToVertexRelation> {
+    let mut v2v = vec![SmallVec::new(); mesh.vertices().len()];
 
     for face in mesh.faces() {
         match face {
             Face::Triangle(f) => {
                 let vertex_indices = &[f.vertices.0, f.vertices.1, f.vertices.2];
                 for i in 0..vertex_indices.len() {
-                    let neighbors = v2v.entry(vertex_indices[i]).or_insert_with(SmallVec::new);
-
                     let neighbor_candidate1 = vertex_indices[(i + 1) % 3];
                     let neighbor_candidate2 = vertex_indices[(i + 2) % 3];
 
-                    if !neighbors.contains(&neighbor_candidate1) {
-                        neighbors.push(neighbor_candidate1)
+                    if !v2v[cast_usize(vertex_indices[i])].contains(&neighbor_candidate1) {
```

Comment:
Similar to above, I am not sure the optimizer can see this, so I'd extract `v2v[cast_usize(vertex_indices[i])]` to a variable.